template/template.go

```diff
@@ -0,0 +1,203 @@
+package template
+
+import (
+	"errors"
+	"fmt"
+	"io"
+	"strings"
+
+	ast "go.expect.digital/mf2/parse"
+)
+
+type execFn func(operand any, opts map[string]any) (string, error)
+
+type Template struct {
+	ast       *ast.AST
+	execFuncs map[string]execFn
+	executer  *executer
+}
+
+type executer struct {
+	wr    io.Writer
+	input map[string]any
+}
+
+func (e *executer) writeString(s string) error {
+	if _, err := e.wr.Write([]byte(s)); err != nil {
+		return fmt.Errorf("write: %w", err)
+	}
+
+	return nil
+}
+
+func newExecuter(wr io.Writer, input map[string]any) *executer {
+	return &executer{wr: wr, input: input}
+}
+
+func New() *Template {
+	return &Template{execFuncs: make(map[string]execFn)}
+}
+
+func Must(t *Template, err error) *Template {
+	if err != nil {
+		panic(err)
+	}
+
+	return t
+}
+
+// AddFunc adds a function to the template's function map.
+func (t *Template) AddFunc(name string, f execFn) {
+	t.execFuncs[name] = f
+}
+
+func (t *Template) Parse(input string) (*Template, error) {
+	ast, err := ast.Parse(input)
+	if err != nil {
+		return nil, fmt.Errorf("%w: %s", ErrSyntax, err.Error())
+	}
+
+	t.ast = &ast
+
+	return t, nil
+}
+
+func (t *Template) Execute(wr io.Writer, input map[string]any) error {
+	if t.ast == nil {
+		return errors.New("AST is nil")
+	}
+
+	t.executer = newExecuter(wr, input)
```

Comment:
Would it make more sense to pass `io.Writer` to downstream functions instead to let the GC free memory when `io.Writer` is not needed anymore?

Current implementation does not release the memory when executor is reused multiple times.

Reply:
Made an alternative, my previous implementation was kinda questionable.